### DISCUSSION TITLE
Support for multi-user environment?

### DISCUSSION THREAD

**Discussion by zDEFz** [TriliumNext/Trilium#1375] [Ideas]
Wouldn't it be great to be able to login with multiple users? Might be beneficial if they have their own little database.
Then, the application could be hosted on potent Servers, right?

What do you think?

**Comment by Engr-AllanG** [comment]
This has been discussed quite a bit here. An answer is given on the [wiki](https://github.com/zadam/trilium/wiki/FAQ). If you search "multi-user" or similar on the closed Issues, you'll see some more discussion on the subject

**Reply by zDEFz** [reply]
Thanks. I haven't had the correct search terms. Have to learn more about this environment :)

**Comment by Engr-AllanG** [comment]
I agree, multi-user support would be great. Personally I am able to achieve this by using multiple containers (proxmox virtualization) each running an instance. I believe you can run two instances on the same machine by changing a few settings in the Trilium config. I chose two separate containers because I have outside access via an nginx reverse proxy, and I thought it was easier to just have two containers with their own IP.

**Reply by meichthys** [reply]
^This is what i do - right down to the external access with NGINX

**Comment by meichthys** [comment]
@zadam Do you know if Trilium would be able to handle numerous (maybe 10 clients) syncing to the same server? As an alternative to multiple users, I'm considering having all users log into the same instance with separate clients. Do you think this could result in sync conflict issues?

**Reply by zadam** [reply]
I haven't personally tested more than 3 clients, but in principle there shouldn't be a problem. Sync conflicts will arise when people will edit the same note at the roughly same time.

**Reply by meichthys** [reply]
@zadam Ok thanks. Are sync conflicts handles automatically? (Like latest edit takes precedence)

I may give it a try.

**Reply by zadam** [reply]
Yes, last edit takes precedence. The conflicted versions might be preserved in the note revisions.

**Reply by meichthys** [reply]
Thanks this sounds feasible!

Is there a way (apart from scripting) to set the entire note tree as read only (or non-editable) by default? 
With multiple users notes can get changed accidentally quite easily (this is a constant issue when using OneNote). Although annoying and not needed in a single user setup, it would be nice to require the users to click an edit button before being able to edit (like we already need to do for large notes or in standard Wiki software).

**Reply by zadam** [reply]
You can put `#readOnly(inheritable)` on the root note, that should do the trick.

**Reply by meichthys** [reply]
Woah! Too easy. Thanks! 
Trilium keeps surprising me with it's versatility.

**Reply by meichthys** [reply]
Does Trilium store any information about the sync clients? From poking around in the SQL Console, i wasn't able to find any identifying information about any of the clients.
I'd like to be able to see which client was the one that created/modified a note.

**Reply by zadam** [reply]
No, there isn't any info stored about other instances.

**Reply by meichthys** [reply]
Ok thank you. I'll look into scripting something. Just need to find out how to get the pc hostname.

**Comment by micmejia** [comment]
@meichthys 

> Does Trilium store any information about the sync clients? From poking around in the SQL Console, i wasn't able to find any identifying information about any of the clients. I'd like to be able to see which client was the one that created/modified a note.

> Ok thank you. I'll look into scripting something. Just need to find out how to get the pc hostname.

I would like to ask if you were able to do these?

I'm thinking of running a single trilium server instance, accessed by about 10 client devices (for a small team). When a note is edited/created, it would be ideal to add information about the particular device where edit came from, e.g. ip address, device hostname and/or startup runtime variable, etc.. tia

**Reply by meichthys** [reply]
@micmejia Yes and No. I was able to get a script working to an extent. If i remember correctly, I was able to get the ip added to a note label, but fetching the hostname of the machine was not possible due to the fact that modern browsers prevent that kind of information leak.  I ended up not implementing it (yet!) since the additional labels cluttered up the 'jump-to' feature with ip addresses (like other labels do):
![image](https://github.com/zadam/trilium/assets/10717998/3a331c3b-ed41-4a0e-9e0b-239f7a19557f)
Also, to properly keep track of all user changes, there would be either many ip_address labels or some metadata would need to be saved in each note. (Perhaps display the last ip address that edited the note at the very top or bottom of the note). This way you could rely on previous versions feature to go back and see which user edited which note. 

All that being said, it should be possible, but to implement it cleanly and in a reliable way would be key (i.e. only save the ip when something actually changes). 

I'll copy my code that i was working on a while back and maybe you can take it from there. It's probably very rough since I don't know javascript (python is my thing) and was just trying to hack something together.

**Reply by meichthys** [reply]
```
// This widget is intended to save the ip address of the machine that last edited a note
// It currently only logs the ip to the browser console but should be able to be extended to
// append the ip to a log note or to add the ip to a note label (I prefer not to do this since it may
// clutter up the labels which show in the 'jump to note' feature.
// -MG

function onClickOpenPath(event) {
    // Check if clicked element is edit button
    if (event.target.tagName === "BUTTON" && event.target.className.includes("bx-edit-alt")) {
        /**
         * Get the user IP throught the webkitRTCPeerConnection
         * @param onNewIP {Function} listener function to expose the IP locally
         * @return undefined
         */
        function getUserIP(onNewIP) { //  onNewIp - your listener function for new IPs
            //compatibility for firefox and chrome
            var myPeerConnection = window.RTCPeerConnection || window.mozRTCPeerConnection || window.webkitRTCPeerConnection;
            var pc = new myPeerConnection({
                iceServers: []
            }),
            noop = function() {},
            localIPs = {},
            ipRegex = /(192(\.[0-9]{1,3}){3}|[a-f0-9]{1,4}(:[a-f0-9]{1,4}){7})/g,
            key;

            function iterateIP(ip) {
                if (!localIPs[ip]) onNewIP(ip);
                localIPs[ip] = true;
            }

             //create a bogus data channel
            pc.createDataChannel("");

            // create offer and set local description
            pc.createOffer().then(function(sdp) {
                sdp.sdp.split('\n').forEach(function(line) {
                    if (line.indexOf('candidate') < 0) return;
                    line.match(ipRegex).forEach(iterateIP);
                });

                pc.setLocalDescription(sdp, noop, noop);
            }).catch(function(reason) {
                // An error occurred, so handle the failure to connect
            });

            //listen for candidate events
            pc.onicecandidate = function(ice) {
                if (!ice || !ice.candidate || !ice.candidate.candidate || !ice.candidate.candidate.match(ipRegex)) return;
                ice.candidate.candidate.match(ipRegex).forEach(iterateIP);
            };
        }

        // Update `lastEditedBy` label 
        getUserIP(function(ip){
            var user = "unknown";
            if (ip == "192.168.0.142") {user = "user1"}
            else if (ip == "192.168.0.111") {user = "user2"}
            else if (ip == "192.168.0.225") {user = "user3"}
            else if (ip == "192.168.0.117") {user = "user4"}
            else if (ip == "192.168.0.143") {user = "user5"}
            else if (ip == "192.168.0.156") {user = "user6"}
            else if (ip == "192.168.0.166") {user = "user7"}
            else {user = "unknown"};
            console.log(document.PathLinkerApi.getActiveContextNote().getAttribute("label", "lastEditedBy"));
            var attribute = document.PathLinkerApi.getActiveContextNote().getAttribute("label", "lastEditedBy");
            if (attribute === undefined){
                attribute = new document.PathLinkerApi.FAttribute(this.froca, );
                attribute.noteId = document.PathLinkerApi.getActiveContextNote().noteId;
                attribute.type = "label";
                attribute.value = user;
            } else {
                attribute.value = user;
            };
        });
    }

}

const TEMPLATE = `
<div style="padding: 10px; border-top: 1px solid var(--main-border-color); contain: none;">
<script>
${onClickOpenPath.toString()}
document.addEventListener('click', onClickOpenPath);
</script>
</div>`;

class LastEditedByUpdaterWidget extends api.NoteContextAwareWidget {
    constructor(...args) {
        super(...args);
        this.balloonEditorCreate = null;
    }
    
    get position() { 
        // higher value means position towards the bottom/right
        return 100; 
    } 

    get parentWidget() { return 'center-pane'; }

    doRender() {
        this.$widget = $(TEMPLATE);
        this.$widget.hide();
        
        // store API in document to access it from callback
        document.PathLinkerApi = api;
        return this.$widget;
    }
}

let widget = new LastEditedByUpdaterWidget();
console.log("Loaded LastEditedByUpdaterWidget");
module.exports = widget;
```

**Reply by meichthys** [reply]
Also be sure to check out some of [my tips and tricks here for using Trilium in a multi-user environment](https://github.com/zadam/trilium/discussions/3921). 
I've been running it with 7 users (7 different sync clients) with almost no issues at all. The key for me was to make all the notes `#readOnly` by default to prevent accidental deletions of entire sub-trees.

If you have any other tips/tricks/ideas, please follow up in that post!

**Reply by micmejia** [reply]
@meichthys Thank you. I'm new to trilium and your detailed response and tips & tricks will help me get up to speed. Especially the tip about making the root `#readOnly`.

In certain cases, I think the requirement to get the IP address/hostname/user id can be further simplified by passing environment variables at startup, e.g. in linux: 

```
TRILIUM_DATA_DIR=/work/notes/tn-mic TRILIUM_PORT=37843 IP=$(hostname -I) USER=mic trilium
```
then read them via javascript's `process.env`:
![2024-01-26_23-11 trilium env vars](https://github.com/zadam/trilium/assets/117283/38d757aa-ec94-43e8-b16f-ea176d398a1c)

idea/next steps workflow:
- set root note as `#readOnly`
- when a user clicks on the **Edit this note** button, javascript will update a note's label, e.g. `editedby`=`ID` and/or `USER` captured in `process.env`
- ideally, if possible the auto-save/sync is disabled. And there will be a button to SAVE/SYNC/COMMIT the changes. This will allow the user to make atomic changes to the note, or cancel the editing if so desired. @meichthys  @zadam is this currently possible? thanks

**Reply by meichthys** [reply]
> In certain cases, I think the requirement to get the IP address/hostname/user id can be further simplified by passing environment variables at startup, e.g. in linux:

Ooo i like that idea. I hadn't thought of that!

**Reply by meichthys** [reply]
It wouldn't be hard to find out, but i'm not sure if note labels are stored in note history. If they aren't then you won't have the 'history' of edits. You may need to store the 'edit information' in the note content.

**Reply by rom1dep** [reply]
Wouldn't it be preferable to have clients identify themselves with a randomly-generated identifier? That could enable some decently reliable session tracking (to label sessions, relate them to users and devices, disconnect them if required, …), and probably some of the legwork is readily available since I can see some session cookies and device IDs being used.

**Reply by meichthys** [reply]
> Wouldn't it be preferable to have clients identify themselves with a randomly-generated identifier? That could enable some decently reliable session tracking (to label sessions, relate them to users and devices, disconnect them if required, …), and probably some of the legwork is readily available since I can see some session cookies and device IDs being used.

Yes this would definitely be ideal but I wasn't able to find any persistent ID that could be used like you mentioned. Could you point us to where we can find this device ID?

**Reply by rom1dep** [reply]
It's a long shot but my browser's dev tools show a bunch of cookies, including what appear to be stable IDs. Anyway, my naive take was pretty much "if we are going to be reinventing cookies to do session tracking, let's have a look at those we've got already".

**Reply by meichthys** [reply]
>  let's have a look at those we've got already\".

Yes for sure. I'll take a look at this

**Comment by micmejia** [comment]
> It wouldn't be hard to find out, but i'm not sure if note labels are stored in note history. If they aren't then you won't have the 'history' of edits. You may need to store the 'edit information' in the note content.

I just did a quick test, and indeed labels are not stored in the note version history.

**Reply by meichthys** [reply]
I guess that leaves us with two options. Either we store the edit data in the note itself, or we store it in the back end. I'm not sure how difficult it would be to store it in the back end since I did not investigate that.

**Reply by micmejia** [reply]
If we store it in the note itself, I think we need to take into consideration the note type, e.g. if it's a markdown, text, python file, js, mermaid diagram, etc -- and use whatever comment syntax the particular note type supports.

I think what's ideal now is to store the edited-by information as another column to whatever database table the version history is being saved. I'll try to study the trilium code when I get the chance; if the core developers can expose an api to do this, or maybe have some special note labels that get stored alongside the note content for each revision -- that would be great.

**Reply by rom1dep** [reply]
are labels versioned? If not, that might hinder traceability quite a lot…

**Reply by micmejia** [reply]
> are labels versioned? If not, that might hinder traceability quite a lot…

based on my quick tests yesterday, labels are not versioned. @zadam would you be able to confirm? thanks
